File: cpp/problem724.cpp
#include <mystd.h>
using namespace std;

/**
 * solution by directly calculating the left-sum and right-sum
 * 
 * The right-sum is initialized by first loop (by calling accumulate).
 * In each step of the (forward) second loop, the left sum is accumulated 
 * gradually, and the right-sum is subtracted gradually.
 * By definition, if we find at some point the left-sum == right-sum, the
 * currecnt index is returned as the answer.
 */
class Solution {
public:
    int pivotIndex(vector<int>& nums) {
    	int rsum = std::accumulate(nums.begin(), nums.end(), 0),
    		lsum = 0;
    	for(auto &i: nums){
    		if( lsum == (rsum -= i) ) return &i - nums.data();
    		lsum += i;
    	}
    	return -1;
    }
};


int main(int argc, char const *argv[]){
	Solution sol;

	// test case in Leetcode Problem
	vector<int> nums = {1, 7, 3, 6, 5, 6};
	cout << sol.pivotIndex( nums ) << endl;
	nums.assign({1,2,3});
	cout << sol.pivotIndex( nums ) << endl;
	return 0;
}

File: java/problem724.java
class Solution {
    public int pivotIndex(int[] nums) {
        int size=nums.length;
        if(size==1){
            return 0;
        }
        for(int i=size-2;i>=0;i--){
            nums[i]+=nums[i+1];
        }
        for(int i=0;i<size;i++){
            if(i!=0&&i!=size-1){
                if(nums[0]-nums[i+1]==nums[i]){
                    return i;
                }
            }else{
                if(i==0&&nums[1]==0){
                    return 0;
                }else if(i==size-1&&nums[0]-nums[i]==0){
                    return i;
                }
            }
        }
        return -1;
    }
}